Migrate/teamcity (#4135)
circleci/circleci-docs: jekyll/_cci2/migrating-from-teamcity.adoc

= Migrating from TeamCity 
:page-layout: classic-docs
:page-liquid:
:source-highlighter: pygments.rb
:icons: font
:toc: macro
:toc-title:
:sectanchors:

This document provides an overview of how to migrate from TeamCity to CircleCI.

== Quick Start

CircleCI has various differences in setup and configuration from TeamCity. For the best experience of trying CircleCI, we recommend getting started with the following:

. **Read (or at least browse) this document** to solidify conceptual differences.
. **Quickly build and run a simple https://circleci.com/docs/2.0/hello-world/#section=getting-started[Hello World job] in any repository**. This configuration demonstrates the simplicity of our YAML setup, and it showcases how to define both an execution environment and the steps to run within it. All jobs on CircleCI share this format.

> . **Once you understand the config, modify the executor and the individual steps to your needs.** For example, a https://circleci.com/docs/2.0/language-ruby/[Rails project] may use a Ruby container and run `bundler` commands. Consult our https://circleci.com/docs/2.0/tutorials/#section=configuration[Tutorials page] for various examples in different languages and frameworks.

Reviewer: ```suggestion
. **Once you understand the config, modify the executor and the individual steps to meet your needs.** For example, a https://circleci.com/docs/2.0/language-ruby/[Rails project] may use a Ruby container and run `bundler` commands. Consult our https://circleci.com/docs/2.0/tutorials/#section=configuration[Tutorials page] for various examples in different languages and frameworks.
```

. **After finishing one job, setup your other jobs on CircleCI and then https://circleci.com/docs/2.0/workflows/[configure a workflow]**. This would essentially migrate your Build Chain.

> . **Next, https://circleci.com/docs/2.0/optimizations/[optimize your jobs] with our advanced features** to see how much faster your builds could run on our platform.

Reviewer: ```suggestion
. **Next, https://circleci.com/docs/2.0/optimizations/[optimize your jobs] with advanced features** to see how much faster your builds could run on the CircleCI platform.
```

> . **Throughout this process, consult our https://circleci.com/docs/2.0/configuration-reference/#section=configuration[Config Reference]** for a list of all available configuration on our platform.

Reviewer: ```suggestion
. **Throughout this process, consult the https://circleci.com/docs/2.0/configuration-reference/#section=configuration[Configuration Reference]** for a list of all available configuration on the CircleCI platform.
```

== Why Migrate to CircleCI?

CircleCI is a modern CI/CD tool aimed at performance and developer productivity. With a flexible system for configuration, you can define complex workflows and jobs for your application's specific needs.

> **Configuration As Code**:: The majority of CircleCI's configuration is done via YAML. It is easy to maintain and trivial to learn, and your CI/CD pipeline can be tracked and versioned like the rest of your source code.

Reviewer: Just wondering what the significance of double "::" is?

Author: Ah oops that's a typo.

Author: @rosieyohannan Oh actually, no it's not a typo. The double colons are description lists in ASCII Doc: https://asciidoctor.org/docs/asciidoc-writers-guide/#description-lists

Reviewer: Oh great! :-D

**Flexibility**:: CircleCI is vendor and tool-agnostic. Our config allows integration with any platform or cloud provider that has an API. See our https://circleci.com/integrations/[integrations page] for existing partnerships. Dependencies and libraries can be installed ad-hoc in your job, and we offer https://circleci.com/docs/2.0/caching/[dependency caching] to speed up subsequent builds.
**Shareable, Reusable Configuration**:: Steps, execution environments, and jobs can be templated, parameterized, shared, and reused across different projects and teams using https://circleci.com/docs/2.0/orb-intro/#section=configuration[CircleCI orbs]. Take advantage of the https://circleci.com/orbs/registry/[numerous integrations and orbs] already written for you.

We have various other features that set our solution apart. https://circleci.com/signup/[Sign up for a free account today] and try us out, or if you are interested in CircleCI for your team, https://circleci.com/talk-to-us/?source-button=MigratingFromGithubActionsDoc[contact our sales team] to set up a trial.

== Concepts

=== High-Level Differences

* CircleCI and TeamCity share the concept of projects.

> * TeamCity's Build Configurations are equivalent to https://circleci.com/docs/2.0/concepts/#jobs[CircleCI's jobs], which are atomic elements of work consisting of many steps or commands to achieve some sort of goal (e.g. run unit tests).

Reviewer: ```suggestion
* TeamCity's Build Configurations are equivalent to https://circleci.com/docs/2.0/concepts/#jobs[CircleCI's jobs], which are atomic elements of work consisting of many steps or commands to achieve a goal (e.g. run unit tests).
```

          requires:
            - Package

# Job definitions assumed
----
|===

[.table.table-striped.cf.table-migrate-mobile]
[cols=1*, options="header", stripes=even]
[cols="100%"]
|===
| TeamCity Build Chain

a|
[source, kotlin]
----
project {
  sequence {
    build(Compile)
    parallel {
        build(Test1)
        build(Test2)
    } 
    build(Package)
    build(Publish)
  }
}

/* BuildType definitions assumed

----
|===

[.table.table-striped.cf.table-migrate-mobile]
[cols=1*, options="header", stripes=even]
[cols="100%"]
|===
| Equivalent CircleCI Workflow

a|
[source, yaml]
----
version: 2.1
workflows:
  build-deploy:
    jobs:
      - Compile
      - Test1:
          requires:
            - Compile
      - Test2:
          requires:
            - Compile
      - Package:
          requires:
            - Test1
            - Test2
      - Publish:
          requires:
            - Package

# Job definitions assumed
----
|===

For more information on CircleCI Concepts, visit our https://circleci.com/docs/2.0/concepts/[Concepts] and https://circleci.com/docs/2.0/pipelines/#section=pipelines[Pipelines] documentation pages.

== Configuration

=== Environment

> TeamCity requires setting up a build agent with the required O.S. & tools installed and a corresponding Build Configuration. In CircleCI, all job configurations have an Executor definition, and CircleCI handles spinning up said agents for you. See our list of https://circleci.com/docs/2.0/executor-types/[available executors].

Reviewer: ```suggestion
TeamCity requires setting up a build agent with the required OS and tools installed and a corresponding Build Configuration. In CircleCI, all job configurations have an Executor definition, and CircleCI handles spinning up said agents for you. See our list of https://circleci.com/docs/2.0/executor-types/[available executors].
```
